cc/gtp_player.cc

```diff
@@ -0,0 +1,323 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "cc/gtp_player.h"
+
+#include <iomanip>
+#include <iostream>
+#include <utility>
+
+#include "absl/strings/str_cat.h"
+#include "absl/strings/str_join.h"
+#include "absl/strings/str_split.h"
+#include "absl/time/clock.h"
+
+namespace minigo {
+
+namespace {
+
+bool CheckArgsExact(absl::string_view cmd, size_t expected_num_args,
+                    const std::vector<absl::string_view>& args) {
+  if (args.size() != expected_num_args) {
+    std::cerr << "ERROR: expected " << expected_num_args
+              << " args for GTP command \"" << cmd << "\", got " << args.size()
+              << " args:";
+    for (const auto& arg : args) {
+      std::cerr << " \"" << arg << "\"";
+    }
+    std::cerr << "\n";
+    return false;
+  }
+  return true;
+}
+
+bool CheckArgsRange(absl::string_view cmd, size_t expected_min_args,
+                    size_t expected_max_args,
+                    const std::vector<absl::string_view>& args) {
+  if (args.size() < expected_min_args || args.size() > expected_max_args) {
+    std::cerr << "ERROR: expected between " << expected_min_args << " and "
+              << expected_max_args << " args for GTP command \"" << cmd
+              << "\", got " << args.size() << " args:";
+    for (const auto& arg : args) {
+      std::cerr << " \"" << arg << "\"";
+    }
+    std::cerr << "\n";
+    return false;
+  }
+  return true;
+}
+
+}  // namespace
+
+GtpPlayer::GtpPlayer(std::unique_ptr<DualNet> network, const Options& options)
+    : MctsPlayer(std::move(network), options),
+      num_readouts_(options.num_readouts) {}
+
+bool GtpPlayer::HandleCmd(const std::string& line) {
+  std::vector<absl::string_view> args =
+      absl::StrSplit(line, absl::ByAnyChar(" \t\r\n"), absl::SkipWhitespace());
+  if (args.empty()) {
+    std::cout << "=" << std::endl;
+    return true;
+  }
+
+  auto cmd = args[0];
+  args.erase(args.begin());
+
+  if (cmd == "quit") {
+    std::cout << "=" << std::endl;
+    return false;
+  }
+
+  auto result = DispatchCmd(cmd, args);
+  if (result == "") {
+    std::cout << "=" << std::endl;
+  } else {
+    std::cout << "= " << result << std::endl;
+  }
+  return true;
+}
+
+absl::Span<MctsNode* const> GtpPlayer::TreeSearch(int batch_size) {
+  auto leaves = MctsPlayer::TreeSearch(batch_size);
+  if (!leaves.empty() && report_search_interval_ != absl::ZeroDuration()) {
+    auto now = absl::Now();
+    if (now - last_report_time_ > report_search_interval_) {
+      last_report_time_ = now;
```

Comment:
consider moving this after the reportsearchstatus. 
On windows some terminals are slow (100 lines per / second) and if half of the duration was taken up by printing the stats might be confusing. but I can also see an arguement that you want output every 1 second and if it only prints every 1.2s that might be weird